packages/augur-core/src/contracts/para/AMMExchange.sol

```diff
@@ -8,236 +8,251 @@ import 'ROOT/ICash.sol';
+            return _setsToBuy.add(_poolYes.sub(_poolConstant.div(_poolNo.add(_setsToBuy))));
+        } else {
+            return _setsToBuy.add(_poolNo.sub(_poolConstant.div(_poolYes.add(_setsToBuy))));
+        }
+    }
+
+    // If you do not have complete sets then you must have more shares than _setsToSell because you will be swapping
+    // some of them to build complete sets.
+    function exitPosition(uint256 _setsToSell) external {
+        (uint256 _noFromUser, uint256 _yesFromUser) = rateExitPosition(_setsToSell);
+
+        // materialize the complete set sale for cash
+        shareTransfer(msg.sender, address(this), _setsToSell, _noFromUser, _yesFromUser);
+        cash.transfer(msg.sender, _cashToBuy);
+    }
+
+    // How many extra shares you need.
+    // Returns (no,yes) or (long,short)
+    function rateExitPosition(uint256 _setsToSell) public view returns (uint256,uint256) {
+        (uint256 _userInvalid, uint256 _userNo, uint256 _userYes) = shareBalances(msg.sender);
+        (uint256 _poolNo, uint256 _poolYes) = yesNoShareBalances(address(this));
+
+        // short circuit if user is closing out their own complete sets
+        if (_userInvalid >= _setsToSell && _userNo >= _setsToSell && _userYes >= _setsToSell) {
+            return (_setsToSell, _setsToSell);
+        }
+
+        // User can only have insufficient invalid shares if they sell them outside of the AMM.
+        require(_userInvalid >= _setsToSell, "AugurCP: You don't have enough invalid tokens to close out for this amount.");
+        require(_userNo > _setsToSell || _userYes > _setsToSell, "AugurCP: You don't have enough YES or NO tokens to close out for this amount.");
+
+        // simulate user swapping enough NO ➡ YES or YES ➡ NO to create setsToSell complete sets
+        uint256 _poolConstant = poolConstant(_poolYes, _poolNo);
+        uint256 _invalidFromUser = _setsToSell;
+        uint256 _noFromUser = 0;
+        uint256 _yesFromUser = 0;
+        if (_userYes > _userNo) {
+            uint256 _noToUser = _setsToSell.sub(_userNo);
+            uint256 _yesToPool = _poolConstant.div(_poolNo.sub(_noToUser)).sub(_poolYes);
+            require(_yesToPool <= _userYes.sub(_setsToSell), "AugurCP: You don't have enough YES tokens to close out for this amount.");
+            _noFromUser = _userNo;
+            _yesFromUser = _yesToPool + _setsToSell;
+        } else {
+            uint256 _yesToUser = _setsToSell.sub(_userYes);
+            uint256 _noToPool = _poolConstant.div(_poolYes.sub(_yesToUser)).sub(_poolNo);
+            require(_noToPool <= _userNo.sub(_setsToSell), "AugurCP: You don't have enough NO tokens to close out for this amount.");
+            _yesFromUser = _userYes;
+            _noFromUser = _noToPool + _setsToSell;
+        }
+
+        return (_noFromUser, _yesFromUser);
+    }
+
+    function swap(uint256 _inputShares, bool _inputYes) external returns (uint256) {
```

Comment:
Will need a parameter to prevent front running here too.  Something like `minimumOutputShares` or something.

Reply:
added